shared-bindings/keypad/KeyMatrix.c

```diff
@@ -0,0 +1,263 @@
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|     def deinit(self) -> None:
+//|         """Stop scanning and release the pins."""
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix_deinit(mp_obj_t self_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_keypad_keymatrix_deinit(self);
+    return MP_ROM_NONE;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(keypad_keymatrix_deinit_obj, keypad_keymatrix_deinit);
+
+//|     def __enter__(self) -> KeyMatrix:
+//|         """No-op used by Context Managers."""
+//|         ...
+//|
+//  Provided by context manager helper.
+
+//|     def __exit__(self) -> None:
+//|         """Automatically deinitializes when exiting a context. See
+//|         :ref:`lifetime-and-contextmanagers` for more info."""
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_keypad_keymatrix_deinit(args[0]);
+    return MP_ROM_NONE;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(keypad_keymatrix___exit___obj, 4, 4, keypad_keymatrix___exit__);
+
+STATIC void check_for_deinit(keypad_keymatrix_obj_t *self) {
+    if (common_hal_keypad_keymatrix_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+//|     num_keys: int
+//|     """The number of keys that are being scanned. (read-only)
+//|     """
+//|
+STATIC mp_obj_t keypad_keymatrix_get_num_keys(mp_obj_t self_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    return MP_OBJ_NEW_SMALL_INT(common_hal_keypad_keymatrix_get_num_keys(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(keypad_keymatrix_get_num_keys_obj, keypad_keymatrix_get_num_keys);
+
+const mp_obj_property_t keypad_keymatrix_num_keys_obj = {
+    .base.type = &mp_type_property,
+    .proxy = {(mp_obj_t)&keypad_keymatrix_get_num_keys_obj,
+              MP_ROM_NONE,
+              MP_ROM_NONE},
+};
+
+//|     def key_num(self, row: int, col: int) -> int:
+//|         """Return the key number for a given row and column.
+//|         The key number is calculated by ``row * len(col_pins) + col``.
+//|         """
+//|         ...
+//|
+STATIC mp_obj_t keypad_keymatrix_key_num(mp_obj_t self_in, mp_obj_t row_in, mp_obj_t col_in) {
+    keypad_keymatrix_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    check_for_deinit(self);
+
+    const mp_uint_t row = (mp_uint_t)mp_arg_validate_int_range(
+        mp_obj_get_int(row_in), 0, (mp_int_t)common_hal_keypad_keymatrix_get_num_rows(self), MP_QSTR_row);
+
+    const mp_int_t col = (mp_uint_t)mp_arg_validate_int_range(
+        mp_obj_get_int(col_in), 0, (mp_int_t)common_hal_keypad_keymatrix_get_num_cols(self), MP_QSTR_col);
+
+    return MP_OBJ_NEW_SMALL_INT(
+        (mp_int_t)common_hal_keypad_keymatrix_key_num(self, row, col));
+}
+MP_DEFINE_CONST_FUN_OBJ_3(keypad_keymatrix_key_num_obj, keypad_keymatrix_key_num);
+
+//|     def pressed(self, key_num: int) -> None:
```

Comment:
Why provide this? A library could do this with the event queue internally.

Reply:
I had requests both for events and for raw key state, both as a fast path. A library would not be a fast path. I think both use cases are valid and would like to cater to both.

Reply:
being able to get events *and* also current key state is really valuable for UI and HID reports

Reply:
Leaving as is

Reply:
What are events useful for? I'm trying to counter having multiple ways of doing something. Furthermore, we should consider Seesaw connected matrices because they should have the same API. (See https://github.com/adafruit/Adafruit_CircuitPython_seesaw/pull/76 )

Reply:
I personally wanted events so I can tell when a key goes down and not have to maintain extra state myself. It's less useful for an HID gadget, where you are just copying over the up/down state to the report, but it is quite useful for a game or some stand-alone appliance. As @ladyada said:

> being able to get events and also current key state is really valuable for UI and HID reports

Reply:
Also for a MIDI keyboard, the events are nice.

Reply:
Having it native is a higher bar than "nice-to-have" though. If we have a consistent current state API, then we can make a library to create events based on it for *all* implementations, not just native ones. A library would do the state tracking for you. One problem with a pure current state API is that it can miss presses. That's why gamepad latches presses and the current state API probably should as well.

How much code size is added by the event queue stuff? Moving it to a library would reduce the overall size.

Reply:
I was presented with multiple use cases. it was a goal to have all of them run quickly. For each case, there is a most convenient way to report and use the state of some keys. The reason I have provided multiple reporting methods natively is to make all these use cases nice and fast. Yes, I could choose one scanning mechanism, and layer others on top, in Python, but that would make the upper layers be slower.